odoo/models.py

```diff
@@ -1075,6 +1075,50 @@ def splittor(rs):
         if not _is_toplevel_call:
             splittor = lambda rs: rs
 
+        # {properties_fname: {record: {property_name: (value, property_type)}}}
+        cache_properties = {}
+
+        def get_property(properties_fname, property_name, record):
+            # FIXME: Only efficient during the _is_toplevel_call == True
```

Comment:
Maybe it is a good time to refactor the api and drop `_is_toplevel_call`
we can move the logic in `for record in splittor(self):` in a local helper function which can be recursively called. So that the `cache_property` can be used by all exported records.

Another thing I am not sure is about the memory usage. When the number of exported record is really large, if it will cause memory issue of this cache. because `splittor` will invalidate env.cache for every 1000 records

Reply:
The main goal of this PR is just to make it work with properties. There is a lot of things we could refactor in import/export. Depends on how easy that change will be.

You want to invalidate to avoid filling the memory. Otherwise, importing a big set of data would just fill-up the memory and just fail at some point.

Reply:
> Maybe it is a good time to refactor the api and drop `_is_toplevel_call`

I don't think that exporting of properties can justify the complete refactoring of the method/export. It will not be used that much (I think). But indeed this method does need a refactor (but not focus on properties field, which is not properly designed with the Field class abstraction IMHO).

> Another thing I am not sure is about the memory usage. When the number of exported record is really large, if it will cause memory issue of this cache. because splittor will invalidate env.cache for every 1000 records

We only save property definitions, so it shouldn't be a problem. Also the recordset invalidation is far from optimal to free memory anyway (if you export compute field, there's a good chance you'll fill up the cache with a bunch of other stuff)